### PR TITLE
Add frame toggling for preview

### PR DESCRIPTION
**Pull request by thorhunter1**
Hi, I've recently found out this project when looking for pixel editing software and it was one of the best free software I've found :) . But it lacks one feature that almost disqualified it from being chosen, namely toggling frames for animation preview. This is almost crucial for implementing several animations for one sprite.

Luckily, I've found out the project was open source and implementation of this feature quite possible and I think others may find it useful, so here it goes:

out

### CODE REVIEW

**Comment by zakarh** [comment]
I really like this. When designing sprites animations I want to focus on specific frames. This makes it possible.

**Comment by thorhunter1** [comment]
I'm glad you like it. After several bug fixes I started making actual sprites with it and it's very useful :).

**Comment by juliandescottes** [comment]
Overall I really like this feature! Thanks a lot for the suggestion and for taking the time to make it work.

Regarding the implementation, we should be able to do something simpler, see suggestions below.

The "frame" objects are described in `js/model/Frame.js` . Here I suggest adding a `visible` property, as well as a toggleVisibility method. Then add a method on the piskelController (eg `toggleFrameAt`) that will toggle the visibility of all frames at the provided index. This is what should be called when performing the toggle action. Then you can remove all the state based on this.toggles in FramesListController. When rendering a frame you simply check if `frame.visible` is true, and set the class accordingly.

To render the animated preview, we can have a "hasVisibleFrameAt" method on the piskel controller and update the animation loop to do something like:

```javascript
      var index = Math.floor(this.elapsedTime / (1000 / this.fps));
      while (!this.piskelController.hasVisibleFrameAt(index)) {
        if (!this.piskelController.hasFrameAt(index)) {
          this.elapsedTime = 0;
          index = 0;
          break;
        }
        index++;
      }
```

I am not sure about the green color to show the frame as "toggled", but we can find another representation that fits the gold/black theme later :)

Another question is "should this apply when exporting" but this can also be handled later on.

**Comment by thorhunter1** [comment]
Initially I took similar approach you proposed but it turns out the history management system is not very forgiving for those who want to add any data to frame model. Basically, any undo/redo causes frames to be recreated and I've given up on trying to add data into piskel state serialization/deserialization.

**Comment by juliandescottes** [comment]
Ah! You probably had some issues with the ArrayBufferSerializer/Deseserializer. Happy to contribute that part of the code. I will try to push a commit that adds support for this, assuming the property and method names I mentioned above.

**Comment by juliandescottes** [comment]
Here is a patch that should support both import/export and history service serializations:
https://gist.githubusercontent.com/juliandescottes/9639d34ecdd717afe79927c98b6d440d/raw/5e4f1786bc53220db2b0a572c8f177ebd651708c/gistfile1.txt

It only assumes a "visible" property on Frame. We can probably come up with cleaner code, but that should be enough to get started. Hopefully you still feel like giving this feature another shot :) Let me know if you have any trouble with the patch

**Comment by juliandescottes** [comment]
Didn't think about it at first, but the other thing needed for the history service is to save the calls to this `toggleFrameAt`.

In PublicPiskelController, adding `this.saveWrap_('toggleFrameAt', true);` should automatically make it so that all calls to the method are part of the history. Meaning they are part of the undo/redo stack, but that's the easiest way to handle them here.

**Comment by thorhunter1** [comment]
I followed your suggestions with the patch you provided and indeed everything seem to work nicely (except for unit tests :)). However, I kept the old implementation for calculating index for preview, because the one you suggested unfortunately has bugs (and I don't think it can be kept simple without running into corner cases). I've also changed color scheme to match gold standard :).

**Comment by juliandescottes** [comment]
Thanks! I will take a look at the unit tests :)

File: src/js/utils/serialization/Serializer.js [hunk 1]
```diff
@@ -21,6 +21,16 @@
       var serializedLayers = piskel.getLayers().map(function (l) {
         return pskl.utils.serialization.Serializer.serializeLayer(l);
       });
+      var frames = pskl.app.piskelController.getLayerAt(0).getFrames();
```
Comment (juliandescottes) [inline src/js/utils/serialization/Serializer.js]: This should be `var frames = piskel.getLayerAt(0).getFrames();` my bad!

**Comment by juliandescottes** [comment]
My tests are failing locally, I must have busted my environment. The PR looks good, if you can just address my comment to make the unit tests pass, I will merge this.

**Comment by juliandescottes** [comment]
Thanks! Merging

**Comment by thorhunter1** [comment]
Ok coming back to this, it looks like there are some corner case bugs where history management toggles all frames instead of chosen ones. Also, creating new layer also doesn't have initial frames toggled properly so it looks like storing toggled status as part of state is not so simple after all :/

**Comment by juliandescottes** [comment]
I'll take a look at this. 

HistoryManagement is operating in 2 modes:
- replay
- snapshot

Basically we store all actions in an array and when you undo we reapply all actions until the previous one.  To avoid redoing everything from the beginning, we regularly store a snapshot of the sprite (every 50 actions).

So when you hit CTRL+Z, we load the last snapshot taken, and then replay all actions. Maybe loading the snapshot fails to reapply the flag correctly, although that's precisely what the serializer work should have fixed.

The layer issue however seems very obvious now that you mention it.

Since our model is Piskel -> Layers -> Frames, the `visible` information becomes hard to maintain as soon as we have more than one layer. This means it should probably move the Piskel model instead, which should just keep the list of hidden indices.

Then we'll need to be careful about which methods also need to update this list (for instance when deleting a frame, moving a frame etc...). But in the long run that will probably be easier to handle than duplicating properly the `visible` information on Frames. 

Sorry for missing this when reviewing initially. Let me know if you want to try to address this, otherwise I'll take it.
